Break goto_symex_state into more manageable bits

romainbrenguier opened the pull request:
The definition of `goto_symex_statet` is long (about 250 lines) and it is difficult to see any structure in it. This PR extracts some bits from it with the aim to make it clearer. It ends up being ~140~ 100 lines which is a bit better but still not great. 
I apologize in advance if this makes it more difficult to rebase some of your branches without really fixing anything, but I find it extremely difficult to understand what this class is doing, so it seems necessary. 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [na] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

romainbrenguier committed:
Extract goto_statet class from goto_symex_statet

This makes the goto_symex_statet class shorter and easier to read.

romainbrenguier committed:
Extract get_original_name from goto_symex_statet

This is not using any fields of the class so can be made static and
moved to renaming_level.h which seems more appropriate for a function
dealing with the renamings.

romainbrenguier committed:
Use auto for iterator types

Iterator types are long and not very informative, it's better to use
auto for these.

romainbrenguier committed:
Use preincrement in for-loop

Postincrement has to make a copy of the iterator, which in some case
can be expensive. It's better to always use preincrement in for loops.

romainbrenguier committed:
Remove typedef used only once

typedef doesn't help clarity if it is used only once and the name of the
type is the same as the variable name.

romainbrenguier committed:
Pass argument by copy to framet constructor

A copy would have been made anyway for initializing the field, we now
can actually avoid it in cases the argument can be moved to the constructor.

romainbrenguier committed:
Extract framet from goto_symex_statet

Its definition is not dependent on goto_symex_statet.
Extracting it makes the goto_symex_statet shorter and easier to read.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -294,4 +280,19 @@ class goto_symex_statet final
   goto_symex_statet(const goto_symex_statet &other) = default;
 };
 
+inline goto_statet goto_state(const goto_symex_statet &s)
```
I don't get it. If there are no dependencies, then we can just move the entire `goto_statet` class out of `goto_symex_statet`. Constructors play a peculiar role, and we ought to make use of that (for example by using `goto_state_list.emplace_back(state);` in the line that this commit also touches?

romainbrenguier replied:
Actually I needed this definition to be able to move the `goto_statet` class out. I added a commit which reintroduce a constructor for `goto_statet`, but after creating another intermediary class, which looks a bit nicer. 
(see the commit: Extract a goto_symex_call_stackt class from goto_symex)

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -30,6 +30,35 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "renaming_level.h"
 #include "symex_target_equation.h"
 
+/// State of the goto program
```
No, this is a state that we have branched to.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -30,6 +30,35 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "symex_target_equation.h"
 
+/// State of the goto program
+class goto_statet
```
Instead of doing the static function introduced in the previous commit, just put this class definition _after_ that of `goto_symex_statet` and be happy?

romainbrenguier replied:
`goto_symex_statet` also has a constructor from `goto_statet` so we would be in the same situation.

tautschnig replied:
Just _define_ the constructors at the end of the file? That is, have all the class definitions first with declarations of the constructors, and then put the definitions of those constructors at the bottom?

tautschnig commented on src/goto-symex/renaming_level.h:
```diff
@@ -86,4 +86,10 @@ struct symex_level2t : public symex_renaming_levelt
+void get_original_name(exprt &expr);
+
+/// Undo all levels of renaming
+void get_original_name(typet &type);
```
Could we please make them static members of `symex_renaming_levelt`? These are only meaningful in that very context.

romainbrenguier replied:
I'm not opposed to it but I don't see why it should be specific to symex_renaming_levelt. What they do seem pretty meaningful to me on its own.

tautschnig replied:
I'm just worried about polluting the global namespace, but not a very big deal.

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -96,8 +96,72 @@ struct framet
   }
 };
 
+/// Intermediary class for encoding state of goto-symex
```
No, the "state of goto-symex" should be in `goto_symext`. This should by the program state at a particular point during symbolic execution.

romainbrenguier replied:
Ok thanks, that's a bit clearer to me now. Although the distinction between `goto_statet` and `goto_symex_statet` is still a bit fuzzy.

tautschnig replied:
A `goto_statet` is the part of a state that can vary across different branches. Or at least most of that (the L1 renaming could also vary, but isn't stored).

tautschnig reviewed (changes requested):
Refactoring sounds like a good idea, with some changes still required.

romainbrenguier committed:
Make goto_symex_statet inherit from goto_statet

All fields of goto_statet are also found in goto_symex_statet.
We can use inheritence two avoid repeating the parts that are common.

Documentation is moved from goto_symex_statet to goto_statet for these
fields.
The debugging function output_propagation_map is also moved to
goto_statet since it can be useful to both.

It makes comparing the two easier; makes goto_symex_statet shorter and
more structured; and make converting one to the other a bit clearer.

smowton commented:
@romainbrenguier @tautschnig I'm inclined to think we can merge `level2_current_names` (in `goto_statet`) with `level2` (in `goto_symex_statet`), what do you think? That would leave `goto_statet::source` the only member unique to `goto_statet`, which AFAICT is only there for debugging, and could be stored in `goto_state_list` instead of directly in `goto_statet`, eliminating the need for `goto_state_baset`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 83fe01cd82b50b3897ed81320ac2fe28739df24a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99442047

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -134,15 +163,9 @@ class goto_symex_call_stackt
-  {
-    call_stack().pop_back();
-  }
+  void pop_frame() { call_stack().pop_back(); }
```
Is this a whitespace-revert merged into the wrong commit?

tautschnig commented:
If we can merge `level2` and `level2_current_names` that would seem good, but I think there are a lot more members left in `goto_symex_statet` that aren't in `goto_statet` (or `goto_state_baset`).

smowton commented:
Yes, that should definitely remain separate (I think we should have per-location state and global state, where the former is called `goto_statet` and `goto_symex_statet` either has or is a `goto_statet`)

romainbrenguier commented:
> @romainbrenguier @tautschnig I'm inclined to think we can merge level2_current_names (in goto_statet) with level2 (in goto_symex_statet), what do you think?

Yes, totally. `level2` is actually just a wrapper for the `current_names` so it makes sense. If we can get `goto_symex_statet` inheriting from `goto_statet`, that would make thing simpler. I will try that.

romainbrenguier committed:
Replace level2_current_names field by level2

symex_level2t is a wrapper for the current_names, so this is appropriate
and this allows us to use directly the methods from this class instead
of having to duplicate the code for them.
So level2_get_variables and level2_current count can be removed.

romainbrenguier committed:
Give default value for goto_statet fields

This ensures we can use default initialization.
This is so we can make goto_symex_statet inherit from goto_statet and
keep the same behaviour for the default constructors.

romainbrenguier commented:
@smowton @tautschnig I implemented the change suggested so that `goto_symex_statet` inherits from `goto_statet`. I've decided to drop one commit which was extracting a call_stack class from this PR, to make it a bit easier to review, and will do that change in a separate PR.

tautschnig reviewed (approved):
Looks good except:
1. Please drop "Replace goto_statet constructor by static function" as it is removed in a later commit and
2. there is a CI failure.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7938cd19f3df65fd8bf63a224b2d469f1cec882a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99744109
